sbi/inference/snre/bnre.py

```diff
@@ -0,0 +1,89 @@
+        NeurIPS 2022. https://arxiv.org/abs/2208.13624
+
+        Args:
+            prior: A probability distribution that expresses prior knowledge about the
+                parameters, e.g. which ranges are meaningful for them. If `None`, the
+                prior must be passed to `.build_posterior()`.
+            classifier: Classifier trained to approximate likelihood ratios. If it is
+                a string, use a pre-configured network of the provided type (one of
+                linear, mlp, resnet). Alternatively, a function that builds a custom
+                neural network can be provided. The function will be called with the
+                first batch of simulations $(\theta, x)$, which can thus be used for shape
+                inference and potentially for z-scoring. It needs to return a PyTorch
+                `nn.Module` implementing the classifier.
+            device: Training device, e.g., "cpu", "cuda" or "cuda:{0, 1, ...}".
+            logging_level: Minimum severity of messages to log. One of the strings
+                INFO, WARNING, DEBUG, ERROR and CRITICAL.
+            summary_writer: A tensorboard `SummaryWriter` to control, among others, log
+                file location (default is `<current working directory>/logs`.)
+            show_progress_bars: Whether to show a progressbar during simulation and
+                sampling.
+            regularization_strength: The multiplicative coefficient applied to the
+                balancing regularizer ($\lambda$)
+        """
+
+        self.regularization_strength = regularization_strength
+        kwargs = del_entries(
+            locals(), entries=("self", "__class__", "regularization_strength")
+        )
+        super().__init__(**kwargs)
+
+    def _loss(self, theta: Tensor, x: Tensor, num_atoms: int) -> Tensor:
+        """Returns the binary cross-entropy loss for the trained classifier.
+
+        The classifier takes as input a $(\theta,x)$ pair. It is trained to predict 1
+        if the pair was sampled from the joint $p(\theta,x)$, and to predict 0 if the
+        pair was sampled from the marginals $p(\theta)p(x)$.
+        """
+
+        assert theta.shape[0] == x.shape[0], "Batch sizes for theta and x must match."
+        batch_size = theta.shape[0]
+
+        logits = self._classifier_logits(theta, x, num_atoms)
+        likelihood = torch.sigmoid(logits).squeeze()
+
+        # Alternating pairs where there is one sampled from the joint and one
+        # sampled from the marginals. The first element is sampled from the
+        # joint p(theta, x) and is labelled 1. The second element is sampled
+        # from the marginals p(theta)p(x) and is labelled 0. And so on.
+        labels = ones(2 * batch_size, device=self._device)  # two atoms
+        labels[1::2] = 0.0
+
+        # Binary cross entropy to learn the likelihood (AALR-specific)
+        bce = nn.BCELoss()(likelihood, labels)
```

Comment:
Is there a reason to not call `super()._loss(theta, x)`?

Reply:
Yes, I need access to the logits to compute the regularizer and the loss function does not return those.